features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp

```diff
@@ -0,0 +1,469 @@
+    return addOrReplaceData(
+        adv_data_type_t::SLAVE_CONNECTION_INTERVAL_RANGE,
+        mbed::make_Span((const uint8_t*)&interval, 4)
+    );
+}
+
+ble_error_t AdvertisingDataBuilder::setServiceData(
+    UUID service,
+    mbed::Span<const uint8_t> data
+) {
+    if (service.getLen() + data.size() > 0xFE) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    adv_data_type_t short_type = adv_data_type_t::SERVICE_DATA_16BIT_ID;
+    adv_data_type_t long_type = adv_data_type_t::SERVICE_DATA_128BIT_ID;
+
+    size_t total_size = data.size() + service.getLen() + 2;
+    size_t old_size = getFieldSize(
+        (service.shortOrLong() == UUID::UUID_TYPE_SHORT) ? short_type : long_type
+    );
+
+    /* if we can't fit the new data do not proceed */
+    if (total_size > data.size() - (_payload_length - old_size)) {
+        return BLE_ERROR_BUFFER_OVERFLOW;
+    }
+
+    /* this will insert only the UUID (and remove old data) */
+    ble_error_t status = setUUIDData(
+        mbed::make_Span(&service, 1),
+        short_type,
+        long_type
+    );
+
+    if (status != BLE_ERROR_NONE) {
+        /* we already checked for size so this must not happen */
+        return BLE_ERROR_INTERNAL_STACK_FAILURE;
+    }
+
+    status = appendData(
+        (service.shortOrLong() == UUID::UUID_TYPE_SHORT) ? short_type : long_type,
+        data
+    );
+
+    if (status != BLE_ERROR_NONE) {
+        return BLE_ERROR_INTERNAL_STACK_FAILURE;
+    }
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t AdvertisingDataBuilder::setLocalServiceList(
+    mbed::Span<const UUID> data,
+    bool complete
+) {
+    adv_data_type_t short_type = complete ?
+        adv_data_type_t::COMPLETE_LIST_16BIT_SERVICE_IDS :
+        adv_data_type_t::INCOMPLETE_LIST_16BIT_SERVICE_IDS;
+
+    adv_data_type_t long_type = complete ?
+        adv_data_type_t::COMPLETE_LIST_128BIT_SERVICE_IDS :
+        adv_data_type_t::INCOMPLETE_LIST_128BIT_SERVICE_IDS;
+
+    return setUUIDData(data, short_type, long_type);
+}
+
+ble_error_t AdvertisingDataBuilder::setRequestedServiceList(
+    mbed::Span<const UUID> data
+) {
+    adv_data_type_t short_type = adv_data_type_t::LIST_16BIT_SOLICITATION_IDS;
+    adv_data_type_t long_type = adv_data_type_t::LIST_128BIT_SOLICITATION_IDS;
+
+    return setUUIDData(data, short_type, long_type);
+}
+
+ble_error_t AdvertisingDataBuilder::getData(
+    mbed::Span<const uint8_t> &data,
+    adv_data_type_t advDataType
+) {
+    uint8_t *field = findField(advDataType);
+    if (field) {
+        uint8_t data_length = field[0] - 1 /* skip type */;
+        data = mbed::make_Span((const uint8_t*)(field + 2 /* skip type and length */), data_length);
+        return BLE_ERROR_NONE;
+    } else {
+        return BLE_ERROR_NOT_FOUND;
+    }
+}
+
+uint8_t* AdvertisingDataBuilder::findField(adv_data_type_t type)
+{
+    /* Scan through advertisement data */
+    for (uint8_t idx = 0; idx < _payload_length; ) {
+        uint8_t fieldType = _buffer[idx + 1];
```

Comment:
Why the +1 on the index

Reply:
Added a documentation bloc explaining advertising data format.